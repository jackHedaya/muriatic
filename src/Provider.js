// @flow // $FlowIssue flow doesn't know hook
import React, { useState, Children } from 'react'
import isPlainObject from './util/isPlainObject'
import StoreContext from './StoreContext'

type Props = {
  Store: Object,
  children: React$Node
}

export default function Provider({ Store, children }: Props) {
  if (!isPlainObject(Store)) {
    throw new Error(
      'react-topdown: Expected the Store to be a PlainObject'
    )
  }

  const [store, setState] = useState(Store)
  const setStore = (state: Object) => {
    if (!isPlainObject(state)) {
      throw new Error(
        'react-topdown: Expected the setStore() argument to be a PlainObject'
      )
    }

    setState({ ...store, ...state })
  }

  return (
    <StoreContext.Provider
      value={{
        store,
        setStore
      }}
    >
      {Children.only(children)}
    </StoreContext.Provider>
  )
}
