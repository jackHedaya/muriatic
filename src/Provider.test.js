import React from 'react'
import Provider from './Provider'
import TestRenderer from 'react-test-renderer'

test(`doesn't work without children`, () => {
  expect(() =>
    TestRenderer.create(<Provider initialState={{}} />)
  ).toThrowErrorMatchingSnapshot()
})

test(`doesn't work without giving initialState`, () => {
  expect(() =>
    TestRenderer.create(
      <Provider>
        <div>elm</div>
      </Provider>
    )
  ).toThrowErrorMatchingSnapshot()
})

const render = initialState =>
  TestRenderer.create(
    <Provider initialState={initialState}>
      <div>elm</div>
    </Provider>
  ).toJSON()

test(`work fine with PlainObject initialState`, () => {
  var tree = render({})
  expect(tree).toMatchSnapshot()

  var tree = render({ name: 'mark' })
  expect(tree).toMatchSnapshot()

  var tree = render({
    chunk: { home: { oneFloor: 'spacious', twoFloor: 'narrow' } },
    array: [1, 2, 3, 4]
  })
  expect(tree).toMatchSnapshot()
})

test(`doesn't work exept PlainObject`,() => {
  expect(() => render(1)).toThrowError('react-topdown: Expected the initialState to be a PlainObject')
  expect(() => render([])).toThrowError('react-topdown: Expected the initialState to be a PlainObject')
  expect(() => render(new Map())).toThrowError('react-topdown: Expected the initialState to be a PlainObject')
  expect(() => render(new Date())).toThrowError('react-topdown: Expected the initialState to be a PlainObject')
  expect(() => render('string')).toThrowError('react-topdown: Expected the initialState to be a PlainObject')
  expect(() => render(true)).toThrowError('react-topdown: Expected the initialState to be a PlainObject')
  expect(() => render(undefined)).toThrowError('react-topdown: Expected the initialState to be a PlainObject')
  expect(() => render(null)).toThrowError('react-topdown: Expected the initialState to be a PlainObject')
  expect(() => render(NaN)).toThrowError('react-topdown: Expected the initialState to be a PlainObject')
})