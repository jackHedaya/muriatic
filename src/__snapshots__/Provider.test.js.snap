// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't work without children 1`] = `"React.Children.only expected to receive a single React element child."`;

exports[`doesn't work without giving Store 1`] = `"react-topdown: Expected the Store to be a PlainObject"`;

exports[`work fine with PlainObject Store 1`] = `
<div>
  elm
</div>
`;

exports[`work fine with PlainObject Store 2`] = `
<div>
  elm
</div>
`;

exports[`work fine with PlainObject Store 3`] = `
<div>
  elm
</div>
`;

exports[`work fine with PlainObject Store 4`] = `
<div>
  elm
</div>
`;
